File: 23.合并k个升序链表.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 *
 * [23] 合并K个升序链表
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
#include<vector>
#include<queue>
using namespace std;
class Solution {
public:
    struct LNGreater{
        bool operator()(ListNode*a,ListNode*b){
            return a->val>b->val;
        }
    };
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        lists.erase(remove(lists.begin(), lists.end(), nullptr), lists.end());
        priority_queue<ListNode*,vector<ListNode*>,LNGreater> a(lists.begin(),lists.end());
        ListNode*ret=new ListNode,*tmp,*rret=ret;
        while(!a.empty()){
            ret->next=a.top();a.pop();
            ret=ret->next;
            if(ret->next!=nullptr)a.push(ret->next);
        }
        return rret->next;
    }
};
// @lc code=end

File: 23.合并-k-个升序链表.py
#
# @lc app=leetcode.cn id=23 lang=python3
#
# [23] 合并 K 个升序链表
#
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
# @lc code=start
from typing import List, Optional
import heapq
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        ListNode.__lt__ = lambda self, other: self.val < other.val
        pq = [i for i in lists if i is not None]
        heapq.heapify(pq)
        head = ListNode()
        tail = head
        while pq:
            tail.next = heapq.heappop(pq)
            tail = tail.next
            if tail.next is not None: heapq.heappush(pq, tail.next)
        tail.next = None
        return head.next
# @lc code=end

